controllers/argocd/openshift/openshift.go

```diff
@@ -92,12 +92,37 @@ func ReconcilerHook(cr *argoprojv1alpha1.ArgoCD, v interface{}, hint string) err
 			}
 			policyRules := getPolicyRuleForApplicationController()
 			policyRules = append(policyRules, clusterRole.Rules...)
+			if !argoprojRulesExists(policyRules) {
+				policyRules = append(policyRules, getPolicyRuleForNonOlmArgoCDApplicationController()...)
+			}
 			o.Rules = policyRules
 		}
 	}
 	return nil
 }
 
+// checks if argoproj.io rules exist in application controller role, if not return false
+func argoprojRulesExists(policies []rbacv1.PolicyRule) (result bool) {
+	result = false
+	for _, policy := range policies {
+		if stringExistsInSlice(policy.APIGroups, "argoproj.io") {
```

Comment:
@saumeya I don't think it is sufficient to only check for the API group here. There could be a situation where the API group is present but permissions were only added for some resources, not others, or some of the verbs are missing

I think it would be better to check whether the exact ruleset is present or not

Reply:
I think these permissions are added by OLM when the operator is installed - so all the rules will exist not just a few verbs/resources etc. What else is the process of adding policy rules?

Reply:
This PR is explicitly dealing with the case where OLM is not present, so we cannot rely on that 
If there is a situation where someone creates a custom role where they include the argoproj group but dont add all the necessary resource/verb permissions correctly, this check will not be able to catch it

Reply:
Please check the logic, if it looks good or needs further improvement.

Reply:
I don't think this is the best approach to do this 
I would suggest something like this instead:

```
func main() {

	a := PolicyRule{
		APIGroups: []string{"grp1", "grp2"},
		Resources: []string{"r1", "r2"},
		Verbs:     []string{"v1"},
	}

	b := PolicyRule{
		APIGroups: []string{"grp2", "grp1"},
		Resources: []string{"r1", "r2"},
		Verbs:     []string{"v1"},
	}

	sort.Strings(b.APIGroups) 
        sort.Strings(b.Resources)
        sort.Strings(b.Verbs)

	fmt.Println(reflect.DeepEqual(a, b))
        // returns true 

}

```

Reply:
I don't think this is the best approach to do this 
I would suggest something like this instead:

```
import (
     ...
     "sort"
     "reflect"
     ...
)

...

func argoprojPermissionsExist (policies []rbacv1.PolicyRule) (result bool) {

        permissionsExist := false
        
	for _, policy := range policies {
		sort.Strings(policy.APIGroups)
		sort.Strings(policy.Resources)
		sort.Strings(policy.Verbs)
		
		if(reflect.DeepEqual(policy, getArgoProjPolicyRule()) {
		    permissionsExist = true
		    break
		}
		
	}
	return permissionsExist
}
```

Reply:
It is just that when I look at the policies added in the role by the operator, the resources are split across rules and in that case the target will not be equal to the policy in the check.

Reply:
I'm not sure I understand the concern
As far as I can tell the only times that resources from the same API group are split across rules is when we explicitly want to set different permissions for different resources (eg retrieve pods but not secrets from "")

I don't see that being the case with argoproj.io resources

Reply:
when i real the role in our operator - it reads like this
`- verbs:
      - '*'
    apiGroups:
      - argoproj.io
    resources:
      - applications
  - verbs:
      - '*'
    apiGroups:
      - argoproj.io
    resources:
      - applicationsets
  - verbs:
      - '*'
    apiGroups:
      - argoproj.io
    resources:
      - appprojects
  - verbs:
      - '*'
    apiGroups:
      - argoproj.io
    resources:
      - argocds
  - verbs:
      - create
      - update
      - patch
      - delete
    apiGroups:
      - argoproj.io
    resources:
      - applications
  - verbs:
      - create
      - update
      - patch
      - delete
    apiGroups:
      - argoproj.io
    resources:
      - applicationsets
  - verbs:
      - create
      - update
      - patch
      - delete
    apiGroups:
      - argoproj.io
    resources:
      - appprojects
  - verbs:
      - create
      - update
      - patch
      - delete
    apiGroups:
      - argoproj.io
    resources:
      - argocds`